src/components/WearablePreview/WearablePreview.controller.ts

```diff
@@ -13,39 +13,62 @@ import mitt, { Emitter } from "mitt"
 const promises = new Map<string, IFuture<any>>()
 const emoteEvents = new Map<MessageEventSource, Emitter<EmoteEvents>>()
 
-window.onmessage = function handleMessage(event: MessageEvent) {
+let isControllerReady = false
+const pendingMessages: MessageEvent[] = []
+
+function processMessage(event: MessageEvent) {
   if (event.data && event.data.type) {
-    switch (event.data.type as PreviewMessageType) {
-      case PreviewMessageType.CONTROLLER_RESPONSE: {
-        const payload = event.data
-          .payload as PreviewMessagePayload<PreviewMessageType.CONTROLLER_RESPONSE>
-        const { id } = payload
-        const promise = promises.get(id)
-        if (promise) {
-          if (payload.ok) {
-            promise.resolve(payload.result)
-          } else if (payload.ok === false) {
-            promise.reject(new Error(payload.error))
+    try {
+      switch (event.data.type as PreviewMessageType) {
+        case PreviewMessageType.CONTROLLER_RESPONSE: {
+          const payload = event.data
+            .payload as PreviewMessagePayload<PreviewMessageType.CONTROLLER_RESPONSE>
+          const { id } = payload
+          const promise = promises.get(id)
+          if (promise) {
+            promises.delete(id)
+            if (payload.ok) {
+              promise.resolve(payload.result)
+            } else if (payload.ok === false) {
+              promise.reject(new Error(payload.error))
+            }
           }
+          break
         }
-        break
-      }
-      case PreviewMessageType.EMOTE_EVENT: {
-        const payload = event.data
-          .payload as PreviewMessagePayload<PreviewMessageType.EMOTE_EVENT>
-        const { type, payload: eventPayload } = payload
-        const events = event.source ? emoteEvents.get(event.source) : null
-        if (events && type) {
-          events.emit(type, eventPayload)
+        case PreviewMessageType.EMOTE_EVENT: {
+          const payload = event.data
+            .payload as PreviewMessagePayload<PreviewMessageType.EMOTE_EVENT>
+          const { type, payload: eventPayload } = payload
+          const events = event.source ? emoteEvents.get(event.source) : null
+          if (events && type) {
+            events.emit(type, eventPayload)
+          }
+          break
+        }
+        case PreviewMessageType.READY: {
+          isControllerReady = true
+          pendingMessages.forEach(processMessage)
+          pendingMessages.length = 0
```

Comment:
Won't we have race collisions if any of the pending message is of type `PreviewMessageType.READY`? or that can't happen?

I am thinking that if any of the pending messages is of that type, it will again fall to this case and execute all pending messages again and after that clean the counter.

Also I think we should do:
```
pendingMessages = []
```

instead of:
`pendingMessages.length = 0`